stdlib/SparseArrays/src/sparsematrix.jl

```diff
@@ -192,58 +186,83 @@ function Base.show(io::IO, ::MIME"text/plain", S::AbstractSparseMatrixCSC)
+    # character can contain up to 4 braille dots in its height (⡇) and up to
+    # 2 braille dots in its width (⠉).
+    if get(io, :limit, true) && (m > 4maxHeight || n > 2maxWidth)
+        s = (m * 2maxWidth ÷ n, n * 4maxHeight ÷ m)
+        if s[1] <= 4maxHeight
+            scaleHeight = s[1]
+            scaleWidth = 2maxWidth
         else
-            print(ioc, Base.undef_ref_str)
+            scaleHeight = 4maxHeight
+            scaleWidth = s[2]
         end
     end
 
-    function _get_cols(from, to)
-        idx = eltype(getcolptr(S))[]
-        c = searchsortedlast(getcolptr(S), from)
-        for i = from:to
-            while i == getcolptr(S)[c+1]
-                c +=1
-            end
-            push!(idx, c)
-        end
-        idx
+    brailleBlocks = split("⠁⠂⠄⡀⠈⠐⠠⢀", "") .|> x -> Int(x[1])
+
+    # `brailleGrid` is used to store the needed braille characters for
+    # the matrix `S`. Each row of the braille pattern to print is stored
+    # in a column of `brailleGrid`.
+    brailleGrid = fill(10240, (scaleWidth - 1) ÷ 2 + 2, (scaleHeight - 1) ÷ 4 + 1)
+    brailleGrid[end, :] .= Int('\n')
+
+    # Let `(i, j)` be an index pair of a matrix `S`. Consider a matrix `B`
+    # of size `height × width`. This method calculates an index pair `(a, b)`,
+    # such that the position of the element `B[a, b]` in `B` corresponds
+    # the best to the position of the element `S[i, j]` in `S`.
+    @inline function _scale_index(i::Integer, j::Integer, S::AbstractSparseMatrixCSC, height::Integer, width::Integer)
+        a = size(S, 1) <= 1 || height <= 1 ? 1.0 : (i - 1) / (size(S, 1) - 1) * (height - 1) + 1
+        b = size(S, 2) <= 1 || width <= 1 ? 1.0 : (j - 1) / (size(S, 2) - 1) * (width - 1) + 1
+        return round.(Int, (a, b))
     end
 
-    rows = displaysize(io)[1] - 4 # -4 from [Prompt, header, newline after elements, new prompt]
-    if !get(io, :limit, false) || rows >= nnz(S) # Will the whole matrix fit when printed?
-        cols = _get_cols(1, nnz(S))
-        padr, padc = ndigits.((maximum(rowvals(S)[1:nnz(S)]), cols[end]))
-        _format_line.(1:nnz(S), cols, padr, padc)
-    else
-        if rows <= 2
-            print(io, "\n  \u22ee")
-            return
+    # Given an index pair `(i, j)` of a matrix, find the corresponding triple
+    # `(k, l, p)` such that the element at `(i, j)` can be found at
+    # position `(k, l)` in the braille grid `brailleGrid` and corresponds
+    # to the 1-dot braille character `brailleBlocks[p]`
+    @inline function _to_braille_grid_space(i::Integer, j::Integer)
+        k = (j - 1) ÷ 2 + 1
+        l = (i - 1) ÷ 4 + 1
+        p = ((j - 1) % 2) * 4 + ((i - 1) % 4 + 1)
+        return (k, l, p)
+    end
+
+    rvals = rowvals(S)
+    @inbounds for j = 1:size(S, 2)
+        for x in nzrange(S, j)
+            si, sj = _scale_index(rvals[x], j, S, scaleHeight, scaleWidth)
+            k, l, p = _to_braille_grid_space(si, sj)
```

Comment:
It seems weird to define functions for these operations since you only call the functions once.  Why not just inline the function bodies into this loop?

Reply:
Also, the `_scale_index` function, once it is inlined, could be simplified (and made faster) by putting
```jl
rowscale = max(1, scaleHeight - 1) / max(1, size(S, 1) - 1)
colscale = max(1, scaleWidth - 1) / max(1, size(S, 2) - 1)
```
outside the loop and then just doing
```jl
si = round(Int, (rvals[x] - 1) * rowscale + 1)
sj = round(Int, (j - 1) * colscale + 1)
```
in the loop body.

Reply:
I have put
```julia
sj = round(Int, (j - 1) * colscale + 1)
```
outside of the inner loop. I have no idea if the compiler is able to optimize this otherwise.